message L4SysProtoMsg {
	// experiment types
	enum ExperimentType {
		GPRFLIP = 10;
		RATFLIP = 15;
		IDCFLIP = 20;
		ALUINSTR = 30;
	}
	// parameters
	required ExperimentType exp_type = 10;
	required int32 instr_offset = 20;
	required int32 bit_offset = 30;

	// results
	// make these optional to reduce overhead for server->client communication
	enum ResultType {
		DONE = 1;
		TIMEOUT = 2;
		TRAP = 3;
		INTR = 4;
		WRONG = 5;
		UNKNOWN = 6;
	}
	// instruction pointer where injection was done
	optional uint32 injection_ip = 40;
	// result type, see above
	optional ResultType resulttype = 50;
	// result data, depending on resulttype (see source code)
	optional uint32 resultdata = 60;
	// generated output
	optional string output = 70;
	// optional textual description of what happened
	optional string details = 80;
}
